### PR TITLE
GEN-482 (fix pyright errors in GenJAX)

### PR DESCRIPTION
**Pull request by littleredcomputer**
- pyright was giving different results in (VS Code, pre-commit) vs. Github action. I adjusted the GitHub action to follow Pylance, as recommended by the action author, which solved that problem. Previously, upright version 1.1.358 was in use; following Pylance moves that version to 1.1.373.
- Sample is replaced with ChoiceMap at all points where type errors follow the use of the former. I understand that ChoiceMap is a narrower concept than Sample, but we do not have enough type infrastructure in GenJAX to support the distinction presently
- the `inference` library is temporarily excluded, as it is not ready for prime-time 
- the handling of static booleans is now handled by `Flag`. The point of `Flag` is to concentrate the logic for making concreteness optimizations in one place. I believe that the systematic use of Flag has corrected a few spots where concreteness optimization had been overlooked, provoking the elision of a few `unmask` calls in the unit test suite. These require special attention to make sure that my conjecture is true.
- Constructing a Mask requires a Flag. It may be better to have the Mask constructor promote bare booleans to Flag types, or, to add a `post_init` assertion that the type is correct. It may be, however, that the user shouldn't be creating their own Masks, and so we can rely on the type checking for library development from now on to enforce correct Mask creation.
- pyright is (perhaps overly) scrupulous about the names of parameters in overridden methods, and doesn't allow overriding `*args` with a sequence of better names. Can't fight city hall, so I have changed this.
- To clear some errors I have resorted to a pattern of `assert isinstance(x, Foo), f"A Foo is required here, not {x}, [because ...]"`, which I think we should adopt since it's harder to figure out what to do if beartype gets involved.
-

### CODE REVIEW

**Comment by sritchie** [comment]
@littleredcomputer can I bug you to add a project description & title? then when we squash merge folks can get context here

**Comment by littleredcomputer** [comment]
> @littleredcomputer can I bug you to add a project description & title? then when we squash merge folks can get context here

You can once I'm finished; then I'll add a resume of the changes and things to look out for. (Pyright is now green, but there are two failing unit tests)

File: .github/workflows/pyright.yml [hunk 1]
```diff
@@ -33,4 +33,4 @@ jobs:
 
       - uses: jakebailey/pyright-action@v1
         with:
-          version: 1.1.358
```
Comment (sritchie) [inline .github/workflows/pyright.yml:36]: interesting, does this pin to whatever pylance is using?

File: notebooks/cookbook/expressivity/conditionals.ipynb [hunk 2]
```diff
@@ -397,9 +397,9 @@
     "key, subkey = jax.random.split(key)\n",
     "tr3 = jitted(subkey, (2.2,))\n",
     "(\n",
-    "    tr1.get_sample()[(\"s\", \"v1\")],\n",
-    "    tr2.get_sample()[(\"s\", \"v2\")],\n",
-    "    tr3.get_sample()[(\"s\", \"v3\")],\n",
+    "    tr1.get_sample()[\"s\", \"v1\"],\n",
```
Comment (sritchie) [inline notebooks/cookbook/expressivity/conditionals.ipynb]: since I don't know the API well yet, are the before/after here equivalent?

File: pyproject.toml [hunk 3]
```diff
@@ -109,7 +109,10 @@ pythonVersion = "3.11"
 venvPath = "."
 venv = ".venv"
 include = ["src", "tests"]
-exclude = ["**/__pycache__"]
+# TODO: inference has been excluded because this code my not be ready for type-correctness.
+#       msgpack dependency is optional, so it doesn't typecheck
```
Comment (sritchie) [inline pyproject.toml:113]: it should typecheck fine if we modify the `poetry install` command to include `--all-extras` as an option.

Looking now, we already do that in `pyright.yml`:

image

so this SHOULD be able to typecheck in CI just fine.

File: pyproject.toml [hunk 4]
```diff
@@ -109,7 +109,10 @@ pythonVersion = "3.11"
...
+#       typecheck doesn't see noxfiles's import of nox
+exclude = ["**/__pycache__", "noxfile.py", "src/genjax/_src/inference/*.py", "src/genjax/_src/core/serialization/msgpack.py"]
```
Comment (sritchie) [inline pyproject.toml:115]: do you want to add `#noqa` surgically, vs excluding these full files?

File: src/genjax/_src/adev/primitives.py [hunk 5]
```diff
@@ -239,7 +242,8 @@ def _inner(probs, primals):
 
 @Pytree.dataclass
 class NormalREPARAM(TailCallADEVPrimitive):
-    def sample(self, key, loc, scale_diag):
+    def sample(self, key, *args):
```
Comment (sritchie) [inline src/genjax/_src/adev/primitives.py:245]: nice, this is all solid.

File: src/genjax/_src/adev/primitives.py [hunk 6]
```diff
@@ -392,8 +396,8 @@ def _inner(alpha, beta):
 class Baseline(ADEVPrimitive):
     prim: ADEVPrimitive
 
-    def sample(self, key, b, *args):
```
Comment (sritchie) [inline src/genjax/_src/adev/primitives.py:395]: odd that this got past ruff before, which usually calls out unused variables correctly.

File: src/genjax/_src/core/generative/choice_map.py [hunk 7]
```diff
@@ -207,8 +205,8 @@ def maybe(cls, flag: Bool | BoolArray, s: "Selection") -> "Selection":
 
 @Pytree.dataclass
 class AllSel(Selection):
-    def check(self) -> Bool | BoolArray:
-        return True
+    def check(self) -> Flag:
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py:206]: praise Flag

File: src/genjax/_src/core/generative/choice_map.py [hunk 8]
```diff
@@ -269,11 +266,11 @@ class StaticSel(Selection):
     addr: ExtendedStaticAddressComponent = Pytree.static()
     s: Selection = Pytree.field()
 
-    def check(self) -> Bool | BoolArray:
-        return False
+    def check(self) -> Flag:
+        return flag(False)
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py:268]: thank you!!

File: src/genjax/_src/core/generative/choice_map.py [hunk 9]
```diff
@@ -353,10 +348,8 @@ class OrSel(Selection):
     s1: Selection
     s2: Selection
 
-    def check(self) -> Bool | BoolArray:
-        check1 = self.s1.check()
-        check2 = self.s2.check()
-        return staged_or(check1, check2)
+    def check(self) -> Flag:
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py:349]: love this API

File: src/genjax/_src/core/generative/choice_map.py [hunk 10]
```diff
@@ -794,10 +789,10 @@ class XorChm(ChoiceMap):
     c1: ChoiceMap
     c2: ChoiceMap
 
-    def get_value(self) -> Optional[Any]:
+    def get_value(self) -> Any:
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py:790]: is this really all we can say about the return type? I guess it is for now until we track type params through.

I'm excited to add what we can salvage of the generics on top of this base.

File: src/genjax/_src/core/generative/core.py [hunk 11]
```diff
@@ -49,7 +46,7 @@
 if TYPE_CHECKING:
     import genjax
 
-_C = TypeVar("_C", bound=Callable)
+_C = TypeVar("_C", bound=Callable[..., Any])
```
Comment (sritchie) [inline src/genjax/_src/core/generative/core.py:49]: nice, I thought I had caught these bare callables

File: src/genjax/_src/core/generative/core.py [hunk 12]
```diff
@@ -1712,7 +1715,7 @@ def update(
     @typecheck
     def assess(
         self,
-        sample: Sample,
+        sample: "genjax.ChoiceMap",
```
Comment (sritchie) [inline src/genjax/_src/core/generative/core.py:1718]: nice, so assess now at the top level takes a `ChoiceMap`? This wasn't ready to generalize to the tells-us-nothing sample I suppose

File: src/genjax/_src/core/interpreters/staging.py [hunk 13]
```diff
@@ -34,40 +39,72 @@
...
+def flag(b: bool | BoolArray):
+    return Flag(b, not isinstance(b, jc.Tracer))
+
+
+@Pytree.dataclass(eq=False)
+class Flag(Pytree):
+    f: bool | BoolArray
+    concrete: bool = Pytree.static()
+
+    def and_(self, f: "Flag") -> "Flag":
```
Comment (sritchie) [inline src/genjax/_src/core/interpreters/staging.py:66]: any interest in overloading `__or__` and `__and__` here so folks can use `&` and `|` or is that too cute?

File: src/genjax/_src/core/interpreters/staging.py [hunk 14]
```diff
@@ -113,13 +149,21 @@ def cached_stage_dynamic(flat_fun, in_avals):
     return typed_jaxpr
 
 
+# This function has been cloned from api_util, since it is not exported from that module
```
Comment (sritchie) [inline src/genjax/_src/core/interpreters/staging.py:161]: you could do this, or could also do `#noqa` on that import... this is probably better.

File: src/genjax/_src/generative_functions/combinators/mask.py [hunk 15]
```diff
@@ -118,9 +118,9 @@ def simulate(
         key: PRNGKey,
         args: tuple,
     ) -> MaskTrace:
-        check, *inner_args = args
-        tr = self.gen_fn.simulate(key, tuple(inner_args))
-        return MaskTrace(self, tr, check)
+        check, inner_args = args[0], args[1:]
```
Comment (sritchie) [inline src/genjax/_src/generative_functions/combinators/mask.py:121]: nice

File: src/genjax/_src/generative_functions/static.py [hunk 16]
```diff
@@ -604,7 +604,7 @@ def assess(
         self,
         sample: ChoiceMap,
         args: tuple,
-    ) -> tuple[ArrayLike, Any]:
```
Comment (sritchie) [inline src/genjax/_src/generative_functions/static.py:607]: nice, was this a bug before?

File: src/genjax/_src/inference/smc.py [hunk 17]
```diff
@@ -665,21 +667,21 @@ def importance(
         key: PRNGKey,
         constraint: Constraint,
         args: tuple,
-    ) -> tuple[Trace, Weight, UpdateProblem]:
+    ) -> tuple[Trace, Weight]:
```
Comment (sritchie) [inline src/genjax/_src/inference/smc.py:670]: this looks like a bigger change that could use some attention from @femtomc

File: src/genjax/pretty.py [hunk 18]
```diff
@@ -47,19 +47,21 @@ def _pytree_handler(node, subtree_renderer):
                     "background_color must be provided if background_pattern is"
                 )
 
-            def wrap_block(block):
+            def wrapper1(block):
```
Comment (sritchie) [inline src/genjax/pretty.py]: are there more descriptive names we can use for these two wrappers?

File: tests/generative_functions/test_switch_combinator.py [hunk 19]
```diff
@@ -222,7 +222,6 @@ def f2():
         # Just select 0 in all branches for simplicity:
         tr = jax.vmap(s.simulate, in_axes=(0, None))(keys, (0, (), ()))
         y = tr.get_choices()["y"]
-        y = y.unmask()
```
Comment (sritchie) [inline tests/generative_functions/test_switch_combinator.py:225]: are masks unwrapped automatically now?

**Review by sritchie — COMMENTED** [review]
First pass!

File: .github/workflows/pyright.yml (hunk 1 above)
Comment (littleredcomputer) [reply .github/workflows/pyright.yml:36]: Yes (apparently with a lag of a day or two)

File: notebooks/cookbook/expressivity/conditionals.ipynb (hunk 2 above)
Comment (littleredcomputer) [reply notebooks/cookbook/expressivity/conditionals.ipynb]: This fix was automatically applied by Ruff: `[ ]` syntax is provided to the `getitem` method as a tuple, so the parens are superfluous

File: tests/generative_functions/test_switch_combinator.py (hunk 19 above)
Comment (littleredcomputer) [reply tests/generative_functions/test_switch_combinator.py:225]: I am hoping that what has happened is that systematic use of `Flag` has uncovered overlooked concreteness optimizations, and that the previous Masks were incorrect. Special attention required here

File: src/genjax/pretty.py (hunk 18 above)
Comment (littleredcomputer) [reply src/genjax/pretty.py]: fixed

File: src/genjax/_src/generative_functions/static.py (hunk 16 above)
Comment (littleredcomputer) [reply src/genjax/_src/generative_functions/static.py:607]: I don't think the previous types were "wrong" but pyright insists that overriders respect the super type here. Taking a second look I have changed to `[Score, Retval]` which is even better

File: src/genjax/_src/core/interpreters/staging.py (hunk 14 above)
Comment (littleredcomputer) [reply src/genjax/_src/core/interpreters/staging.py:161]: ruff deletes `#noqa` on imports :(

File: src/genjax/_src/core/interpreters/staging.py (hunk 13 above)
Comment (littleredcomputer) [reply src/genjax/_src/core/interpreters/staging.py:66]: I could be talked into this

File: src/genjax/_src/core/generative/core.py (hunk 12 above)
Comment (littleredcomputer) [reply src/genjax/_src/core/generative/core.py:1718]: I am hoping we can reconsider the type question by building on a green state

File: src/genjax/_src/adev/primitives.py (hunk 6 above)
Comment (littleredcomputer) [reply src/genjax/_src/adev/primitives.py:395]: this was an odd case! It was used to drop an arg, so maybe that was enough of an "effect"?

File: pyproject.toml (hunk 4 above)
Comment (littleredcomputer) [reply pyproject.toml:115]: I think that #noqa is the wrong choice here, since it suggests that what's being done is intentional. The inference library doesn't really seem ready to publish to me, but maybe we need a new Issue to cover the expansion of uprightness to this module

File: pyproject.toml (hunk 3 above)
Comment (littleredcomputer) [reply pyproject.toml:113]: I will try this

File: pyproject.toml (hunk 4 above)
Comment (sritchie) [reply pyproject.toml:115]: issue sounds great

File: pyproject.toml (hunk 3 above)
Comment (littleredcomputer) [reply pyproject.toml:113]: this didn't work (the poetry-install hook does not appear to run, despite the preconditions in the documentation having been met)

File: notebooks/cookbook/basics/choicemap_creation_selection.ipynb [hunk 20]
```diff
@@ -324,7 +324,7 @@
     "\n",
     "\n",
     "trace = jax.jit(model.simulate)(key, (0.5,))\n",
-    "trace.get_sample()[\"s\", \"v2\"]"
+    "trace.get_sample()[\"s\", \"v1\"]"
```
Comment (sritchie) [inline notebooks/cookbook/basics/choicemap_creation_selection.ipynb:327]: notice this change, was this a bug?

File: src/genjax/_src/core/generative/choice_map.py [hunk 21]
```diff
@@ -329,10 +325,8 @@ class AndSel(Selection):
     s1: Selection
     s2: Selection
 
-    def check(self) -> Bool | BoolArray:
-        check1 = self.s1.check()
-        check2 = self.s2.check()
-        return staged_and(check1, check2)
+    def check(self) -> Flag:
+        return self.s1.check().and_(self.s2.check())
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py:328]: if you do `__and__`, you can still call it directly as a method if you don't want the explicit `&`

**Review by sritchie — APPROVED** [review]

File: notebooks/cookbook/basics/choicemap_creation_selection.ipynb (hunk 20 above)
Comment (littleredcomputer) [reply notebooks/cookbook/basics/choicemap_creation_selection.ipynb:327]: PRNG instability? The trace does contain the one but not the other. I looked in the past to see if there was evidence that something else had changed but this build has been broken for a while so I didn't find any evidence to the contrary.
